### ISSUE TITLE
Errors on Windows

### ISSUE THREAD

**Issue by saraedum** [Normaliz/Normaliz#383]
The build of Normaliz for Windows on conda-forge reports several segfaults and other errors during the tests, see https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=377874&view=logs&jobId=a70f640f-cc53-5cd3-6cdc-236a1aa90802&j=a70f640f-cc53-5cd3-6cdc-236a1aa90802&t=f5d15007-a01c-5ad8-c9ce-4d519d3b275f

For example, we get:
```
/bin/sh: line 1:  7288 Segmentation fault      /d/bld/normaliz_1631912451318/work/source/normaliz.exe -c --inv test-/medium > /dev/null
make[3]: *** [test-/medium.out] Error 139
```

But there are also errors such as:
```
/d/bld/normaliz_1631912451318/work/source/normaliz.exe -JoeFv test-clopt/whow --all-files --NoExtRaysOutput  --OutputDir=Dummy
WARNING: deprecated option -e is ignored.
Fatal error: Cannot write to output file.
This should not happen, please contact the developers!
FatalException caught... exiting.
make[3]: *** [test-clopt/whow.out] Error 2
```

**Comment by saraedum** [comment]
@isuruf our build does not detect these errors. Normaliz is published with these errors but probably not functional.

**Comment by w-bruns** [comment]
I am quite sure that I have seen the first error resulting from running medium.in (and other input files) before, namely in the parallelized build for Mac OS in our GitHub workflow: the reason is (as far as I could find out) that Normaliz asks for more threads than the environment is willing to allow. For this reason `./github/worklflows/.github-ci-build_check.sh` contains   

       export NORMPARA=-x=1 ## more is not possible on GitHub Mac

 Not even 2 threads were allowed (I haven't tried for a while).  Please try again with a limit on the number of threads.

I know that OutputDir does not work under Windows. About 5 years ago I inserted this option on Matthias Köppe's request. But he never used it, and most likely nobody else ever did. I will try to fix it.

**Comment by saraedum** [comment]
> I am quite sure that I have seen the first error resulting from running medium.in (and other input files) before, namely in the parallelized build for Mac OS in our GitHub workflow: the reason is (as far as I could find out) that Normaliz asks for more threads than the environment is willing to allow. For this reason `./github/worklflows/.github-ci-build_check.sh` contains
> 
> ```
>    export NORMPARA=-x=1 ## more is not possible on GitHub Mac
> ```
> 
> Not even 2 threads were allowed (I haven't tried for a while). Please try again with a limit on the number of threads.

These tests run an Azure and not on GitHub. The underlying infrastructure is likely similary though. I am sure that Azure (and GitHub Actions) allow several threads on macOS. I am running several workflows that run multi-threaded tests and there is no problem there.

If I understand you correctly, this is only an issue in CI? I can test on a mac to see if it works there without this parameter. Or is it known to work on macOS otherwise?

**Comment by saraedum** [comment]
@w-bruns I don't have a Windows running at the moment. Maybe you could try to see whether these Windows problems come from the CI by running with conda-forge's Normaliz.

Which features do the Windows binaries that you built locally have enabled? The ones on conda-forge says:
```
++ ./configure --prefix=/d/bld/normaliz_1642478281995/_h_env/Library --with-nauty=/d/bld/normaliz_1642478281995/_h_env/Library --with-gmp=/d/bld/normaliz_1642478281995/_h_env/Library
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-pc-mingw64
checking host system type... x86_64-pc-mingw64
checking how to print strings... printf
checking for style of include used by make... GNU
checking for gcc... clang.exe
checking whether the C compiler works... yes
checking for C compiler default output file name... a.exe
checking for suffix of executables... .exe
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... no
checking whether clang.exe accepts -g... yes
checking for clang.exe option to accept ISO C89... none needed
checking whether clang.exe understands -c and -o together... yes
checking dependency style of clang.exe... none
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for non-GNU ld... lld-link
checking if the linker (lld-link) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... llvm-nm
checking the name lister (llvm-nm) interface... BSD nm
checking whether ln -s works... no, using cp -pR
checking the maximum length of command line arguments... 8192
checking how to convert x86_64-pc-mingw64 file names to x86_64-pc-mingw64 format... func_convert_file_msys_to_w32
checking how to convert x86_64-pc-mingw64 file names to toolchain format... func_convert_file_msys_to_w32
checking for lld-link option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... (cached) pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... func_cygming_dll_for_implib_fallback
checking for archiver @FILE support... @
checking for strip... no
checking for ranlib... llvm-ranlib
checking command to parse llvm-nm output from clang.exe object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... mt
checking if mt is a manifest tool... yes
checking how to run the C preprocessor... clang.exe -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking whether nauty supports TLS... no
checking whether e-antic headers and library are available... no
checking that generated files are newer than configure... done
configure: creating ./config.status
```

**Comment by w-bruns** [comment]
I agree that my explanation for the problem is dubious. It could very well be a clang vs. gcc (and Intel) problem.

In fact, we have run into a problem with clang 11 on Linux. See https://github.com/Normaliz/Normaliz/issues/374. Our GitHub Linux build with clang and full parallelization explicitly asks for clang 12. It is very well possible that we had run into the same problem in our MacOS builds and guessed a false reason.

So far we have not succeeded in the Windows compilation under MSYS MinGW64 with autotools because the linker runs into severe problems. For this reason I can't present a configure log file. The version we make with source/Makefile.classic includes OpenMP, nauty and hash-library, but no other optional library. My main goal for Windows compilation is a fully static normaliz.exe. It would of course be nice to have further optional packages in it.

Unfortunately I have no Mac, but I know enough people with Macs. I am asking one of them to test our Mac binary 3.9.1 produced for Mac OS.

**Comment by w-bruns** [comment]
On GitHub both Apple and Homebrew use clang++ 13. Therefore I have removed the limit 1 for the parallelization, and the problem with medium.in has disappeared. Therefore I think that my interpretation was wrong.

However, in the two parallelized Mac OS tests a new problem came up: wrong results for hickerson-16 at one place. Like medium.in it is a quite old input file. I have never had a problem with it, and I am at a loss.

**Comment by w-bruns** [comment]
I have tried to take care of the problems that have shown up in the computations by inserting a final extra reduction step in the Hilbert basis computations. In the worst case these are taotological.

The Mac buids on GitHub no go through without a thread limit. 

Please test the current master in your environment. 

I plan a new release for next week.

**Comment by w-bruns** [comment]
With clang-11 on Ubuntu 20.04 I sporadically get the message

```
Assertion failure at kmp_dispatch.cpp(1226): victim.
OMP: Error #13: Assertion failure at kmp_dispatch.cpp(1226).
OMP: Hint Please submit a bug report with this message, compile and run commands used, and machine configuration info including native compiler and operating system versions. Faster response will be obtained by including all program sources. For information on submitting this issue, please see https://bugs.llvm.org/.
```

I have not yet tried to locate the source in the Normaliz code. I would assume that it is a clang problem.

**Comment by w-bruns** [comment]
It seems that the cause of the problem has beenn found. See lhttps://github.com/Normaliz/Normaliz/issues/385.
